### PR TITLE
feat: add delete and update for simple indexer

### PR DESCRIPTION
**Pull request by Yongxuanzhang**

### CODE REVIEW

File: jinahub/indexers/SimpleIndexer/simple_indexer.py [hunk 1]
```diff
@@ -97,6 +97,45 @@ def search(
...
+        if docs is None:
+            return
+        traversal_paths = parameters.get(
+            'traversal_paths', self.default_traversal_paths
+        )
+        flat_docs = docs.traverse_flat(traversal_paths)
+        delete_docs_ids = set(flat_docs.get_attributes('id'))
+        for idx in reversed(range(len(self._docs))):
+            if self._docs[idx].id in delete_docs_ids:
+                del self._docs[idx]
```
Comment (numb3r3) [inline jinahub/indexers/SimpleIndexer/simple_indexer.py:116]: I think the  `__delete__` method in `DAM` can use `doc.id` directly
```suggestion
        for doc in flat_docs:
            if doc.id is None: continue
            deleted_doc = self._docs.get(doc.id, None)
            if deleted_doc:
                del self._docs[doc.id]
        
```

File: jinahub/indexers/SimpleIndexer/simple_indexer.py [hunk 2]
```diff
@@ -97,6 +97,45 @@ def search(
...
+            'traversal_paths', self.default_traversal_paths
+        )
+        flat_docs = docs.traverse_flat(traversal_paths)
+
+        update_docs_ids = set(flat_docs.get_attributes('id'))
+        for idx in reversed(range(len(self._docs))):
+            if self._docs[idx].id in update_docs_ids:
+                update_doc = flat_docs.get_attributes_with_docs(self._docs[idx].id)[1][0]
+                del self._docs[idx]
+                self._docs.append(update_doc)
```
Comment (numb3r3) [inline jinahub/indexers/SimpleIndexer/simple_indexer.py:137]:
```suggestion
        for doc in flat_docs:
            if doc.id is not None:
                self._docs[doc.id] = doc
            else:
                self._docs.append(doc)
        
```

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: jinahub/indexers/SimpleIndexer/simple_indexer.py (hunk 1 above)
Comment (alaeddine-13) [reply jinahub/indexers/SimpleIndexer/simple_indexer.py:116]: can't you do:
```python
        delete_docs_ids = set(flat_docs.get_attributes('id'))
        for idx in delete_docs_ids:
            if idx in self._docs:
                del self._docs[idx]
```
?

File: jinahub/indexers/SimpleIndexer/simple_indexer.py (hunk 2 above)
Comment (alaeddine-13) [reply jinahub/indexers/SimpleIndexer/simple_indexer.py:137]: I do agree with @numb3r3 's suggestion.
simple tweak: maybe we should not accept updating docs with `None` ids or with ids that don't exists in the DAM

File: jinahub/indexers/SimpleIndexer/simple_indexer.py (hunk 1 above)
Comment (numb3r3) [reply jinahub/indexers/SimpleIndexer/simple_indexer.py:116]: oh, looks much better

Comment (Yongxuanzhang) [reply jinahub/indexers/SimpleIndexer/simple_indexer.py:116]: what is the time complexity of `idx in self._docs` ? is it O(1)? If so then it is better. @alaeddine-13

**Review by numb3r3 — APPROVED** [review]
LGTM

**Pull request merged**